### PR TITLE
Implement thoughtsense and apply to Empathic Sense ancestry feature

### PR DESCRIPTION
**Pull request by farling42**
This is the implementation suggested by https://github.com/foundryvtt/pf2e/issues/15111

It also updates the "Empathic Sense" ancestry feature for the Kashrishi race to use this new sense out to 15 ft (even behind walls).

The new sense will only detect Actor types of: character, npc, familiar.

It is best tested by giving the PC with "Empathic Sense" the "Deaf" condition (otherwise they can hear it).

This can also be applied to the Cauthooj (in Monster Core) - "The cauthooj senses all non-mindless creatures at the listed range."

Note that Rhu-Chalik (in Monster Core) have this as a precise sense.

### CODE REVIEW

File: src/module/canvas/perception/modes.ts [hunk 1]
```diff
@@ -152,12 +152,57 @@ class DetectionModeTremorPF2e extends DetectionModeTremor {
     }
 }
 
+class ThoughtsDetectionMode extends DetectionMode {
+    constructor() {
+        super({
+            id: "thoughtsense",
+            label: "PF2E.Actor.Creature.Sense.Type.Thoughts",
```
Comment (simonward) [inline src/module/canvas/perception/modes.ts:159]:
```suggestion
            label: "PF2E.Actor.Creature.Sense.Type.Thoughtsense",
```

**Comment by stwlam** [comment]
I think we'd want a distinct visual effect for this so it can be identified to a user.

**Comment by farling42** [comment]
Updates change the colour to green.
It fixes the case of the "id" so that "Thoughtsense" works when Rules-Based Vision is disabled.
Set correct localization.

**Comment by farling42** [comment]
Cauthooj creature wrongly had tremorsense when it should have thoughtsense.
Other creatures in Monster Core already have thoughtsense configured on their Actors.

File: src/module/canvas/perception/modes.ts [hunk 2]
```diff
@@ -152,12 +152,57 @@ class DetectionModeTremorPF2e extends DetectionModeTremor {
     }
 }
 
+class ThoughtsDetectionMode extends DetectionMode {
+    constructor() {
+        super({
+            id: "senseThoughts",
+            label: "PF2E.Actor.Creature.Sense.Type.Thoughtsense",
+            type: DetectionMode.DETECTION_TYPES.OTHER,
+            walls: false,
```
Comment (stwlam) [inline src/module/canvas/perception/modes.ts:161]:
```suggestion
```

File: src/module/canvas/perception/modes.ts [hunk 3]
```diff
@@ -152,12 +152,57 @@ class DetectionModeTremorPF2e extends DetectionModeTremor {
...
+    protected override _canDetect(visionSource: PointVisionSourcePF2e, target: PlaceableObject): boolean {
+        // Not if the target isn't a token
+        if (!(target instanceof TokenPF2e)) return false;
+
+        // Not if the token is GM-hidden
+        if (target.document.hidden) return false;
+
+        // Detection only works on creatures
+        if (!target.actor.isOfType("character") && !target.actor.isOfType("npc") && !target.actor.isOfType("familiar"))
+            return false;
```
Comment (stwlam) [inline src/module/canvas/perception/modes.ts:185]:
```suggestion
        if (!target.actor.isOfType("creature")) return false;
```

File: src/module/canvas/perception/modes.ts [hunk 4]
```diff
@@ -152,12 +152,57 @@ class DetectionModeTremorPF2e extends DetectionModeTremor {
...
+
+        // Detection cails on mindless creatures
+        if (target.actor.system?.traits?.value?.includes("mindless")) return false;
```
Comment (stwlam) [inline src/module/canvas/perception/modes.ts:188]:
```suggestion
        if (target.actor.system.traits.value.includes("mindless")) return false;
```

This also needs to be blocked by mental immunity.

File: src/module/scene/helpers.ts [hunk 5]
```diff
@@ -96,6 +96,11 @@ function computeSightAndDetectionForRBV(token: TokenDocumentPF2e | PrototypeToke
         const range = scene?.flags.pf2e.hearingRange ?? null;
         token.detectionModes.push({ id: "hearing", enabled: true, range });
     }
+
+    const thsense = actor.perception.senses.get("thoughtsense");
+    if (thsense) {
+        token.detectionModes.push({ id: "senseThoughts", enabled: true, range: thsense.range });
+    }
```
Comment (stwlam) [inline src/module/scene/helpers.ts:103]:
```suggestion
    const thoughtsense = actor.perception.senses.get("thoughtsense");
    if (thoughtsense) {
        token.detectionModes.push({ id: "senseThoughts", enabled: true, range: thoughtsense.range });
    }
```
This shouldn't be added if it's of vague acuity.

**Review by stwlam — CHANGES_REQUESTED** [review]

**Comment by stwlam** [comment]
You might consider rebasing in the future instead of merging master on top.

File: src/module/canvas/perception/modes.ts [hunk 6]
```diff
@@ -152,12 +152,57 @@ class DetectionModeTremorPF2e extends DetectionModeTremor {
...
+
+    static override getDetectionFilter(): OutlineOverlayFilter {
+        const filter = (this._detectionFilter ??= OutlineOverlayFilter.create({
+            wave: true,
+            knockout: false,
+            outlineColor: [0, 1, 0, 1],
+        }));
+        filter.thickness = 1;
+        return filter;
+    }
```
Comment (stwlam) [inline src/module/canvas/perception/modes.ts:174]: This needs to depend on acuity, showing the token in full if the sense is precise.

File: src/module/canvas/perception/modes.ts (hunk 4 above)
Comment (farling42) [reply src/module/canvas/perception/modes.ts]: Test for `target.actor.attributes.immunities.includes("mental")` added

File: src/module/scene/helpers.ts (hunk 5 above)
Comment (farling42) [reply src/module/scene/helpers.ts]: Test for `thoughtsense.acuity !== "vague"` added

File: src/module/canvas/perception/modes.ts (hunk 6 above)
Comment (farling42) [reply src/module/canvas/perception/modes.ts:173]: getDetectionFilter() is a static function, so it isn't provided with the specific token providing the vision.

**Pull request closed without merging**